### PR TITLE
Fix a few minor bugs to get KeyboardWalk to work reliably.

### PR DESCRIPTION
**Pull request by Claegtun**
This branch is mainly to collect a few small fixes for a few small bugs on the NUSense firmware. After this, KeyboardWalk should be mostly reliable.

Such bugs were:
- NUSense hitting a hard fault (like a segmentation fault), possibly because of a race-condition with the CDC callback,
- a servo missing frames because its dirty servo-state was ignored since the dirty-flag was reset after some delay,
- and the firmware getting stuck on a blocking function with an indefinite timeout for the IMU.

Such fixes are:
- to disable the OTG_HS IRQ when the ring-buffer for the USB-communications is being handled lest there may be a race-condition*,
- to reset the dirty-flag of the servo-state straight away before a write-instruction is sent lest a ServoTarget is ignored as the flag is reset afterwards,
- to change the packet timeout-timer to time single bytes rather than a whole packet being decoded lest the Dynamixel loop is throttled too much†,
- to make the IMU's polling timeout lest the firmware gets bricked at that one spot‡,
- and to change the `memcpy`'s in the CDC callback to be for-loops since the former throws away the volatile qualifier.

\* This is rough fix since the fault itself was [an imprecise or asynchronous bus-fault](https://www.keil.com/appnotes/files/apnt209.pdf). [Disabling an internal buffer](https://www.beningo.com/how-to-debug-a-hard-fault-on-an-arm-cortex-m/) apparently makes it easier to troubleshoot. However, disabling the IRQ seems to work well. No hard-fault comes up when the robot is standing for about half an hour. Furthermore, too much time should not be spent since StreamReactor may have a better approach. Interestingly, this bug only happened with the firmware was fully optimised.
† I am not sure about this since it does allow a more accurate timing of missing bytes, but it may never time out if there is a consistent enough stream of garbage. However, Dex's independent watchdog idea may help this.
‡ We may want to make a more thorough DMA approach; see [this task](https://nubots.atlassian.net/browse/HW-68?atlOrigin=eyJpIjoiNDEyNGFhNTg5MjYyNDE5NDk3NWRiMjM1MTY3ZWI2MGYiLCJwIjoiaiJ9).

### CODE REVIEW

**Review by ysims — APPROVED** [review]
LGTM but make sure these ideas are documented somewhere, Jira or GitHub issue

File: NUSense/USB_DEVICE/App/usbd_cdc_if.c [hunk 1]
```diff
@@ -266,14 +266,24 @@ static int8_t CDC_Receive_HS(uint8_t* Buf, uint32_t *Len)
...
   if (rx_buffer.size < RX_BUF_SIZE) {
-      // If the max buffer size is exceeded, wrap around using 2 memcpy calls
+      // If the max buffer size is exceeded, wrap around using two copies.
       if (rx_buffer.back + *Len > RX_BUF_SIZE) {
-          memcpy(&rx_buffer.data[rx_buffer.back], &Buf[0], RX_BUF_SIZE - rx_buffer.back);
-          memcpy(&rx_buffer.data[0], &Buf[RX_BUF_SIZE - rx_buffer.back], rx_buffer.back + *Len - RX_BUF_SIZE);
+          // Use a dumb for-loop since memcpy throws the volatile qualifier away which apparently 
+          // can lead to undefined behaviour.
+          for (int i = 0; i < RX_BUF_SIZE - rx_buffer.back; i++) {
+            rx_buffer.data[rx_buffer.back + i] = Buf[i];
```
Comment (JohanneMontano) [inline NUSense/USB_DEVICE/App/usbd_cdc_if.c:274]: Just making sure this doesn't need a wrap around? i.e., `% RX_BUF_SIZE`. Could get another hardfault if not handled properly

File: NUSense/USB_DEVICE/App/usbd_cdc_if.c [hunk 2]
```diff
@@ -266,14 +266,24 @@ static int8_t CDC_Receive_HS(uint8_t* Buf, uint32_t *Len)
...
+          }
       }
-      // If not then 1 memcpy call should suffice
+      // If not, then one copy should be enough.
       else {
-          memcpy(&rx_buffer.data[rx_buffer.back], &Buf[0], *Len);
+          // Use a dumb for-loop since memcpy throws the volatile qualifier away which apparently 
+          // can lead to undefined behaviour.
+          for (int i = 0; i < *Len; i++) {
+            rx_buffer.data[rx_buffer.back + i] = Buf[i];
```
Comment (JohanneMontano) [inline NUSense/USB_DEVICE/App/usbd_cdc_if.c:285]: Same as above

**Comment by JohanneMontano** [comment]
Should try and send a massive pb message with all 20 servos, just to make sure we're not missing irqs from the read callback. The memcpy was to "guarantee" that the copy routine was fast enough to exit that callback as quick as it could so no interrupts are missed.

**Pull request merged**